dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -154,118 +137,73 @@ class AsyncRequestHandler : public AsyncRequestHandlerBase
         }
         else
         {
+            setCall(reader->getClientContext()->c_call());
             startAsyncRead();
         }
     }
 
-    void processWaitBatchRead(bool ok)
+    void processWaitRead(bool ok)
     {
-        if (ok)
-            ++read_packet_index;
-
-        if (!ok || read_packet_index == batch_packet_count || packets[read_packet_index - 1]->hasError())
-            processReceivedData();
-        else
-            reader->read(packets[read_packet_index], thisAsUnaryCallback());
-    }
-
-    void processWaitFinish()
-    {
-        if (likely(finish_status.ok()))
-            closeConnection("");
-        else
+        if (!ok)
         {
-            String done_msg = fmt::format("Exchange receiver meet error : {}", finish_status.error_message());
-            String log_msg = fmt::format("Finish fail. err code: {}, err msg: {}, retry time {}", finish_status.error_code(), finish_status.error_message(), retry_times);
-            retryOrDone(std::move(done_msg), log_msg);
+            stage = AsyncRequestStage::WAIT_FINISH;
+            reader->finish(finish_status, asGRPCKickTag());
+            return;
         }
-    }
 
-    void startAsyncRead()
-    {
-        stage = AsyncRequestStage::WAIT_BATCH_READ;
-        read_packet_index = 0;
-        received_packet_index = 0;
-        reader->read(packets[read_packet_index], thisAsUnaryCallback());
-    }
-
-    void processReceivedData()
-    {
-        LOG_TRACE(log, "Received {} packets.", read_packet_index);
-
-        String err_info;
-        if (read_packet_index > 0)
-            has_data = true;
-
-        if (auto error_message = getErrorFromPackets(); unlikely(!error_message.empty()))
+        has_data = true;
+        // Check memory overflow error
+        packet->recomputeTrackedMem();
+        if unlikely (packet->hasError())
         {
-            closeConnection(fmt::format("Exchange receiver meet error : {}", error_message));
+            closeConnection(fmt::format("Exchange receiver meet error : {}", packet->error()));
             return;
         }
 
-        GRPCReceiveQueueRes res = sendPackets();
-
-        if (!checkResultAfterSendingPackets(res))
+        stage = AsyncRequestStage::WAIT_PUSH_TO_QUEUE;
```

Comment:
Maybe we should set stage when `res == MPMCQueueResult::FULL`, so moving it to L170 is better.

Reply:
The data race will happen. If it returns `MPMCQueueResult::FULL`, another grpc thread may get this structure and do some work.

Reply:
> The data race will happen. If it returns `MPMCQueueResult::FULL`, another grpc thread may get this structure and do some work.

okk, leave a comment here?

Reply:
Done.